pkg/lib/authenticationflow/declarative/util.go

```diff
@@ -143,176 +142,215 @@ func flowRootObjectForSignupLoginFlow(deps *authflow.Dependencies, flowReference
-			fallthrough
+			options = useAuthenticationOptionAddRecoveryCodes(options, isOptional, userHasSomeSecondaryAuthenticators)
 		case config.AuthenticationFlowAuthenticationPrimaryPassword:
-			fallthrough
+			options = useAuthenticationOptionAddPrimaryPassword(options)
 		case config.AuthenticationFlowAuthenticationPrimaryPasskey:
-			fallthrough
-		case config.AuthenticationFlowAuthenticationSecondaryPassword:
-			fallthrough
-		case config.AuthenticationFlowAuthenticationSecondaryTOTP:
-			err := byUser(branch.Authentication)
+			options, err = useAuthenticationOptionAddPasskey(options, deps, userID, userHasPrimaryPasskeys)
 			if err != nil {
 				return nil, err
 			}
+		case config.AuthenticationFlowAuthenticationSecondaryPassword:
+			options = useAuthenticationOptionAddSecondaryPassword(options, isOptional, userHasSomeSecondaryAuthenticators)
+		case config.AuthenticationFlowAuthenticationSecondaryTOTP:
+			options = useAuthenticationOptionAddTOTP(options, isOptional, userHasSomeSecondaryAuthenticators)
 		case config.AuthenticationFlowAuthenticationPrimaryOOBOTPEmail:
 			fallthrough
 		case config.AuthenticationFlowAuthenticationPrimaryOOBOTPSMS:
-			fallthrough
-		case config.AuthenticationFlowAuthenticationSecondaryOOBOTPEmail:
-			fallthrough
-		case config.AuthenticationFlowAuthenticationSecondaryOOBOTPSMS:
 			if targetStepName := branch.TargetStep; targetStepName != "" {
-				err := byTarget(branch.Authentication, targetStepName)
+				info, err := findIdentity(targetStepName)
 				if err != nil {
 					return nil, err
 				}
+
+				options = useAuthenticationOptionAddPrimaryOOBOTPOfIdentity(options, deps, branch.Authentication, info)
 			} else {
-				err := byUser(branch.Authentication)
-				if err != nil {
-					return nil, err
-				}
+				options = useAuthenticationOptionAddPrimaryOOBOTPOfAllIdentities(options, deps, branch.Authentication, identities)
 			}
+		case config.AuthenticationFlowAuthenticationSecondaryOOBOTPEmail:
+			fallthrough
+		case config.AuthenticationFlowAuthenticationSecondaryOOBOTPSMS:
+			options = useAuthenticationOptionAddSecondaryOOBOTP(options, deps, branch.Authentication, authenticators)
 		}
 	}
 
 	return options, nil
 }
 
-func getAuthenticationOptions(deps *authflow.Dependencies, userID string, as []*authenticator.Info, recoveryCodes []*mfa.RecoveryCode, allAllowed []config.AuthenticationFlowAuthentication) (allUsable []UseAuthenticationOption, err error) {
-	addPrimaryPassword := func() {
-		count := len(as)
-		allUsable = append(allUsable, NewUseAuthenticationOptionPassword(
-			config.AuthenticationFlowAuthenticationPrimaryPassword,
-			count,
+func useAuthenticationOptionAddRecoveryCodes(options []UseAuthenticationOption, isOptional bool, userHasSomeSecondaryAuthenticators bool) []UseAuthenticationOption {
+	shouldAdd := false
+	switch {
+	case !isOptional:
+		// We always add recovery_code even though the end-user does not actually has any.
```

Comment:
Is it because we want the frontend to show the option even the user doesn't have recovery code?

Reply:
But seems some other options such as passkey only shows if user has one. Will it be a bit inconsistent?
Actually, is "always show some options" a ui logic which should not affect the api?

Reply:
Let me update the code so that passkey is always shown when it is enabled.

> a ui logic which should not affect the api?

Authui now rely solely on the information of authflow to render. Authui cannot "peak" the configuration to something extra. Since primary_password is always shown, to be consistent with this behavior, I made all other options always shown as well (except passkey, which I will fix it).

Reply:
Added two commits to address this comment.